File: 2033-minimum-operations-to-make-a-uni-value-grid/solution.py
class Solution:
    def minOperations(self, grid: list[list[int]], x: int) -> int:
        # Flatten the 2D grid into a 1D list
        nums = []
        for row in grid:
            nums.extend(row)
        
        # Check if it's possible to make all elements equal
        # All elements must have the same remainder when divided by x
        remainder = nums[0] % x
        for val in nums:
            if val % x != remainder:
                return -1
        
        # Sort the numbers to find the median
        nums.sort()
        
        # The target value that minimizes the sum of absolute differences is the median
        n = len(nums)
        median = nums[n // 2]
        
        # Calculate total operations to transform every element into the median
        total_operations = 0
        for val in nums:
            total_operations += abs(val - median) // x
            
        return total_operations

File: 2033-minimum-operations-to-make-a-uni-value-grid/solution.cpp
#include <vector>
#include <algorithm>
#include <cmath>

using namespace std;

class Solution {
public:
    int minOperations(vector<vector<int>>& grid, int x) {
        int m = grid.size();
        int n = grid[0].size();
        int totalElements = m * n;
        vector<int> nums;
        nums.reserve(totalElements);

        // Flatten the grid and check if all elements have the same remainder mod x
        int remainder = grid[0][0] % x;
        for (int i = 0; i < m; ++i) {
            for (int j = 0; j < n; ++j) {
                if (grid[i][j] % x != remainder) {
                    return -1;
                }
                nums.push_back(grid[i][j]);
            }
        }

        // Sort to find the median
        sort(nums.begin(), nums.end());
        
        // The median is the optimal value to minimize absolute differences
        int median = nums[totalElements / 2];
        int operations = 0;

        for (int val : nums) {
            operations += abs(val - median) / x;
        }

        return operations;
    }
};